pilot/pkg/serviceregistry/kube/controller/ambientindex_external.go

```diff
@@ -374,3 +451,168 @@ func (c *Controller) getControllerWorkloadEntries(ns string) []*apiv1alpha3.Work
 	}
 	return allWorkloadEntries
 }
+
+// name format: <cluster>/<group>/<kind>/<namespace>/<name></section-name>
+// if the WorkloadEntry is inlined in the ServiceEntry, we may need section name. caller should use generateServiceEntryUID
+func (c *Controller) generateWorkloadEntryUID(wkEntryNamespace, wkEntryName string) string {
+	return c.clusterID.String() + "/networking.istio.io/WorkloadEntry/" + wkEntryNamespace + "/" + wkEntryName
+}
+
+// name format: <cluster>/<group>/<kind>/<namespace>/<name></section-name>
+// section name should be the WE address, which needs to be stable across SE updates (it is assumed WE addresses are unique)
+func (c *Controller) generateServiceEntryUID(svcEntryNamespace, svcEntryName, addr string) string {
+	return c.clusterID.String() + "/networking.istio.io/ServiceEntry/" + svcEntryNamespace + "/" + svcEntryName + "/" + addr
+}
+
+func (c *Controller) cleanupOldServiceEntryVips(svc *model.Service, updates map[model.ConfigKey]struct{}) {
+	a := c.ambientIndex
+	nsName := types.NamespacedName{
+		Name:      svc.Attributes.ServiceEntryName,
+		Namespace: svc.Attributes.ServiceEntryNamespace,
+	}
+
+	// cleanup any old WorkloadEntries generated from this ServiceEntry (`ServiceEntry.endpoints`)
+	if oldServiceEntry, f := a.servicesMap[nsName]; f && oldServiceEntry.Attributes.ServiceEntry != nil {
+		for _, oldWe := range oldServiceEntry.Attributes.ServiceEntry.Endpoints {
+			oldUID := c.generateServiceEntryUID(nsName.Namespace, nsName.Name, oldWe.Address)
+			we, found := a.byUID[oldUID]
+			if found {
+				updates[model.ConfigKey{Kind: kind.Address, Name: we.ResourceName()}] = struct{}{}
+				for _, networkAddr := range networkAddressFromWorkload(we) {
+					delete(a.byWorkloadEntry, networkAddr)
+				}
+				delete(a.byUID, oldUID)
+			}
+		}
+
+		// cleanup any old ServiceEntry VIPs on pods
+		vipsToPorts := getVIPsFromServiceEntry(oldServiceEntry, nil)
+		for vip := range vipsToPorts {
+			// we also need to remove the VIP from any pod that has it (e.g. from service entry `workloadSelector`)
+			for nwAddr := range a.byPod {
```

Comment:
the unit tests actually already covered this scenario, and this code can be removed since pod cleanup is handled during re-construction of the pod workloads on SE update. removed dead code in https://github.com/istio/istio/pull/45621/commits/0eb683ec5198ba57db504923fef37d8dd31774e1